File: lib/repositories/factory_repository.dart
import 'package:supabase_flutter/supabase_flutter.dart';
import '../models/factory_model.dart';

class FactoryRepository {
  final SupabaseClient _client;

  FactoryRepository(this._client);

  Future<List<FactoryModel>> getFactories({String? query}) async {
    var request = _client.from('factories').select('*, factory_spsi_types(*), factory_prices(*)');
    if (query != null && query.trim().isNotEmpty) {
      final q = query.trim();
      request = request.or('name.ilike.%$q%,address.ilike.%$q%');
    }
    final response = await request.order('name', ascending: true);
    return (response as List).map((e) => FactoryModel.fromJson(e)).toList();
  }

  Future<FactoryModel> getFactory(String id) async {
    final response = await _client
        .from('factories')
        .select('*, factory_spsi_types(*), factory_prices(*)')
        .eq('id', id)
        .single();
    return FactoryModel.fromJson(response);
  }

  Future<void> createFactory(FactoryModel factory) async {
    await _client.from('factories').insert(factory.toJson());
    await _replaceTypes(factory);
    await _replacePrices(factory);
  }

  Future<void> updateFactory(FactoryModel factory) async {
    final data = factory.toJson()
      ..remove('id')
      ..remove('created_at');
    await _client.from('factories').update(data).eq('id', factory.id);
    await _replaceTypes(factory);
    await _replacePrices(factory);
  }

  Future<void> deleteFactory(String id) async {
    await _client.from('factory_prices').delete().eq('factory_id', id);
    await _client.from('factory_spsi_types').delete().eq('factory_id', id);
    await _client.from('factories').delete().eq('id', id);
  }

  Future<void> _replaceTypes(FactoryModel factory) async {
    final existing = await _client.from('factory_spsi_types').select('id, name').eq('factory_id', factory.id);
    final existingByName = <String, String>{};
    for (final t in (existing as List)) {
      existingByName[t['name']?.toString().trim().toUpperCase() ?? ''] = t['id'].toString();
    }
    for (final item in factory.spsiTypes) {
      if (item.name.trim().isEmpty) continue;
      final upperName = item.name.trim().toUpperCase();
      final data = item.toJson()..remove('id');
      if (existingByName.containsKey(upperName)) {
        await _client.from('factory_spsi_types').update(data).eq('id', existingByName[upperName]!);
      } else {
        data['factory_id'] = factory.id;
        await _client.from('factory_spsi_types').insert(data);
      }
    }
  }

  Future<void> _replacePrices(FactoryModel factory) async {
    await _client.from('factory_prices').delete().eq('factory_id', factory.id);
    final rows = factory.prices
        .where((item) => item.name.isNotEmpty && item.price > 0)
        .map((item) => item.toJson()..remove('id'))
        .toList();
    if (rows.isNotEmpty) {
      await _client.from('factory_prices').insert(rows);
    }
  }
}

File: app_express/src/repositories/factoryRepository.js
const { assertNoError } = require('./base');
const { env } = require('../config/env');

function db(supabase) {
  return typeof supabase.schema === 'function' ? supabase.schema(env.supabaseSchema) : supabase;
}

function text(value) {
  return String(value || '').trim().toUpperCase();
}

function array(value) {
  return value === undefined || value === null ? [] : Array.isArray(value) ? value : [value];
}

function serializeFactory(body) {
  return { name: text(body.name), address: text(body.address) || null, updated_at: new Date().toISOString() };
}

function serializeSpsi(body, factoryId) {
  const names = array(body.spsi_name);
  const modes = array(body.spsi_mode);
  const amounts = array(body.spsi_amount);
  return names.map((name, index) => ({
    factory_id: factoryId,
    name: text(name),
    calculation_mode: modes[index] === 'FIX' ? 'FIX' : 'PER_KG',
    amount: Number(amounts[index] || 0),
    updated_at: new Date().toISOString()
  })).filter((item) => item.name);
}

function serializePrices(body, factoryId) {
  const names = array(body.price_name);
  const prices = array(body.price_value);
  const defaults = array(body.price_default);
  return names.map((name, index) => ({
    factory_id: factoryId,
    name: text(name),
    price: Number(prices[index] || 0),
    is_default: defaults[index] === '1' || defaults[index] === true,
    updated_at: new Date().toISOString()
  })).filter((item) => item.name && item.price > 0);
}

async function listFactories(supabase) {
  return assertNoError(await db(supabase).from('factories').select('*, factory_spsi_types(*), factory_prices(*)').order('name'));
}

async function getFactory(supabase, id) {
  return assertNoError(await db(supabase).from('factories').select('*, factory_spsi_types(*), factory_prices(*)').eq('id', id).single());
}

async function saveTypes(supabase, factoryId, body) {
  const formTypes = serializeSpsi(body, factoryId);
  const existing = assertNoError(
    await db(supabase).from('factory_spsi_types').select('id, name').eq('factory_id', factoryId)
  );
  const existingByName = {};
  existing.forEach(t => { existingByName[t.name] = t.id; });

  for (const t of formTypes) {
    if (existingByName[t.name]) {
      assertNoError(await db(supabase).from('factory_spsi_types').update(t).eq('id', existingByName[t.name]));
    } else {
      assertNoError(await db(supabase).from('factory_spsi_types').insert(t));
    }
  }
}

async function savePrices(supabase, factoryId, body) {
  const formPrices = serializePrices(body, factoryId);
  const existing = assertNoError(
    await db(supabase).from('factory_prices').select('id, name').eq('factory_id', factoryId)
  );
  const existingByName = {};
  existing.forEach(p => { existingByName[p.name] = p.id; });

  for (const p of formPrices) {
    if (existingByName[p.name]) {
      assertNoError(await db(supabase).from('factory_prices').update(p).eq('id', existingByName[p.name]));
    } else {
      assertNoError(await db(supabase).from('factory_prices').insert(p));
    }
  }

  // Hapus harga yang tidak ada di form
  const toRemove = existing.filter(p => !formPrices.some(fp => fp.name === p.name)).map(p => p.id);
  if (toRemove.length) {
    assertNoError(await db(supabase).from('factory_prices').delete().in('id', toRemove));
  }
}

async function createFactory(supabase, body) {
  const factory = serializeFactory(body);
  if (!factory.name) throw new Error('Nama Pabrik wajib diisi.');
  const created = assertNoError(await db(supabase).from('factories').insert(factory).select().single());
  await saveTypes(supabase, created.id, body);
  await savePrices(supabase, created.id, body);
  return getFactory(supabase, created.id);
}

async function updateFactory(supabase, id, body) {
  const factory = serializeFactory(body);
  if (!factory.name) throw new Error('Nama Pabrik wajib diisi.');
  assertNoError(await db(supabase).from('factories').update(factory).eq('id', id));
  await saveTypes(supabase, id, body);
  await savePrices(supabase, id, body);
  return getFactory(supabase, id);
}

async function deleteFactory(supabase, id) {
  assertNoError(await db(supabase).from('factories').delete().eq('id', id));
}

async function getDefaultPrice(supabase, factoryId) {
  if (!factoryId) return null;
  const { data, error } = await supabase
    .from('factory_prices')
    .select('price')
    .eq('factory_id', factoryId)
    .eq('is_default', true)
    .maybeSingle();
  if (error) throw error;
  return data?.price || null;
}

module.exports = { createFactory, deleteFactory, getDefaultPrice, getFactory, listFactories, updateFactory };
